lib/orbf/rules_engine/services/contract_service.rb

```diff
@@ -0,0 +1,104 @@
+require "json"
+
+module Orbf
+  module RulesEngine
+    class ContractService
+      def initialize(program_id:, all_event_sql_view_id:, dhis2_connection:, calendar:)
+        @program_id = program_id
+        @all_event_sql_view_id = all_event_sql_view_id
+        @dhis2_connection = dhis2_connection
+        @calendar = calendar
+      end
+
+      def program
+        @program ||= begin
+          dhis2_connection.programs.find(@program_id,
+                                         fields:
+                                                 "id,name,programStages[programStageDataElements[dataElement[id,name,code,optionSet[id,name,code,options[id,name]]]]",
+                                         paging: false)
+        end
+      end
+
+      def mappings
+        @mappings ||= begin
+          data_elements = program.program_stages.flat_map do |ps|
+            ps["program_stage_data_elements"].map do |psde|
+              psde["data_element"]
+            end
+          end
+
+          data_elements.each_with_object({}) { |de, mappings| mappings[de["id"]] = de }
```

Comment:
Don't think it will ever be a (performance) problem, but you could do the flat_map and ~inject~ each_with_object in one `each_with_object` and save a loop